File: Beecrowd_Parte4/Quadtree_II_1725.java
import java.util.Scanner;

public class Quadtree_II_1725_ERRO {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        // Ler o tamanho da imagem
        String line1 = scanner.nextLine();
        int width = Integer.parseInt(line1.split(" ")[2]);

        String line2 = scanner.nextLine();
        int height = Integer.parseInt(line2.split(" ")[2]);

        // Ler os dados hexadecimais
        StringBuilder hexData = new StringBuilder();
        while (scanner.hasNextLine()) {
            String line = scanner.nextLine().trim();
            if (line.equals("};")) {
                break;
            }
            if (line.startsWith("0x")) {
                hexData.append(line.substring(2).replaceAll("\\s+", ""));
            }
        }

        scanner.close();

        // Processar os dados hexadecimais
        char[][] pixels = new char[width][height];
        int index = 0;

        for (int y = 0; y < height; y++) {
            for (int x = 0; x < width; x++) {
                // Converter o caractere hexadecimal para um número inteiro
                char hexChar = hexData.charAt(index);
                int hexValue = Character.digit(hexChar, 16);

                // Determinar o valor do pixel (B ou W)
                int bitPosition = 7 - (index % 8); // Posição do bit dentro do byte
                int bitValue = (hexValue >> bitPosition) & 1;

                pixels[x][y] = (bitValue == 0) ? 'B' : 'W';

                index++;
            }
        }

        // Construir o quadtree
        StringBuilder quadtree = new StringBuilder();
        buildQuadtree(pixels, quadtree, 0, 0, width);

        // Imprimir resultado
        System.out.println(width);
        System.out.println(quadtree.toString());
    }

    private static void buildQuadtree(char[][] pixels, StringBuilder quadtree, int x, int y, int size) {
        if (size == 1) {
            quadtree.append(pixels[x][y]);
            return;
        }

        char firstPixel = pixels[x][y];
        boolean same = true;

        // Verificar se todos os pixels na área são iguais
        for (int dy = 0; dy < size; dy++) {
            for (int dx = 0; dx < size; dx++) {
                if (pixels[x + dx][y + dy] != firstPixel) {
                    same = false;
                    break;
                }
            }
            if (!same) break;
        }

        if (same) {
            quadtree.append(firstPixel);
        } else {
            int newSize = size / 2;
            quadtree.append('Q');
            buildQuadtree(pixels, quadtree, x, y, newSize); // Top-left
            buildQuadtree(pixels, quadtree, x + newSize, y, newSize); // Top-right
            buildQuadtree(pixels, quadtree, x, y + newSize, newSize); // Bottom-left
            buildQuadtree(pixels, quadtree, x + newSize, y + newSize, newSize); // Bottom-right
        }
    }
}

File: Beecrowd_Parte4/Quadtree_II_1725.c
#include <stdio.h>

#define MAX_SIZE 512

// Função para construir o quadtree recursivamente
void build_quadtree(char pixels[MAX_SIZE][MAX_SIZE], char *quadtree, int x, int y, int size) {
    if (size == 1) {
        *quadtree = pixels[x][y];
        return;
    }

    char first_pixel = pixels[x][y];
    int same = 1;

    // Verifica se todos os pixels na área são iguais
    for (int dx = 0; dx < size; dx++) {
        for (int dy = 0; dy < size; dy++) {
            if (pixels[x + dx][y + dy] != first_pixel) {
                same = 0;
                break;
            }
        }
        if (!same) break;
    }

    if (same) {
        *quadtree = first_pixel;
    } else {
        *quadtree = 'Q';
        int new_size = size / 2;
        build_quadtree(pixels, quadtree + 1, x, y, new_size);  // Top-left
        build_quadtree(pixels, quadtree + 1 + new_size * new_size, x + new_size, y, new_size);  // Top-right
        build_quadtree(pixels, quadtree + 1 + 2 * new_size * new_size, x, y + new_size, new_size);  // Bottom-left
        build_quadtree(pixels, quadtree + 1 + 3 * new_size * new_size, x + new_size, y + new_size, new_size);  // Bottom-right
    }
}

int main() {
    int N;
    char pixels[MAX_SIZE][MAX_SIZE];

    // Ler tamanho da imagem
    scanf("#define quadtree_width %d\n", &N);
    scanf("#define quadtree_height %d\n", &N);

    // Pular a linha de definição dos bits
    scanf("static char quadtree_bits[] = {\n");

    // Ler os pixels em formato hexadecimal
    int index = 0;
    char hexValue;
    while (scanf("0x%c%c,", &hexValue, &hexValue) == 2) {
        int byte = hexValue >= 'a' ? hexValue - 'a' + 10 : hexValue - '0';
        for (int i = 7; i >= 0; i--) {
            pixels[index / N][index % N] = (byte >> i) & 1 ? 'W' : 'B';
            index++;
        }
    }

    // Pular o caractere de fechamento do array
    getchar();

    // Construir o quadtree
    char quadtree[MAX_SIZE * MAX_SIZE / 3];  // Estimativa de tamanho máximo
    build_quadtree(pixels, quadtree, 0, 0, N);

    // Imprimir o resultado
    printf("%d\n", N);
    printf("%s\n", quadtree);

    return 0;
}
